internal/server/polling/server.go

```diff
@@ -0,0 +1,107 @@
+package polling
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/weaveworks/tf-controller/internal/git/provider"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const DefaultPollingInterval = time.Second * 30
+
+type Server struct {
+	log             logr.Logger
+	clusterClient   client.Client
+	configMapRef    client.ObjectKey
+	pollingInterval time.Duration
+}
+
+func New(options ...Option) (*Server, error) {
+	server := &Server{log: logr.Discard()}
+
+	for _, opt := range options {
+		if err := opt(server); err != nil {
+			return nil, err
+		}
+	}
+
+	return server, nil
+}
+
+func (s *Server) Start(ctx context.Context) error {
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		// Read the config in each iteration. The idea behind this decision is to
+		// allow the user to change the list of resources without the need of
+		// restart of the pod.
+		// It can be a bit smarter like using a time.Ticker and refresh config
+		// periodically.
+		config, err := s.readConfig(ctx)
+		if err != nil {
+			return err
+		}
+
+		secret, err := s.getSecret(ctx, client.ObjectKey{
+			Namespace: config.SecretNamespace,
+			Name:      config.SecretName,
+		})
+		if err != nil {
+			s.log.Error(err, "failed to get secret")
+		}
+
+		for _, resource := range config.Resources {
+			if err := s.poll(ctx, resource, secret); err != nil {
+				s.log.Error(err, "failed to check pull request")
+			}
+		}
+
+		time.Sleep(s.pollingInterval)
+	}
+}
+
+func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret *corev1.Secret) error {
+	if secret == nil {
+		return fmt.Errorf("secret is not defined")
+	}
+
+	tf, err := s.getTerraform(ctx, resource)
```

Comment:
I'm not sure that getTerraform and getSecret are doing much for you here, they don't add anything other than declaring the object for you. Is there any prospect of getting these things any way other than via the client? I think you can just save a couple of func definitions (and future readers looking them up) and do

```
var tf Terraform
err := s.clusterClient.Get(ctx, resource, &tf)
```

(Side note: at present, if a secret or terraform isn't there, you'll get an error something like `failed to get Terraform object: unable to get Terraform: not found`, which repeats itself)

Reply:
It adds an extra layer because we may have to change the query to handle if the source is v1, v2, alphasometingelse. Right now `Terraform` is a `v1alpha2` and `Source` is `v1beta2`, it can be `v1`. So in the future this function can grow, extracting it into a function ensures no one tries to add all those into one function and let it grow 80-100+ lines.